contrib/format-msgpack/src/main/java/org/apache/drill/exec/store/msgpack/MsgpackReader.java

```diff
@@ -0,0 +1,192 @@
+import org.apache.drill.exec.store.msgpack.valuewriter.AbstractValueWriter;
+import org.apache.drill.exec.store.msgpack.valuewriter.ArrayValueWriter;
+import org.apache.drill.exec.store.msgpack.valuewriter.BinaryValueWriter;
+import org.apache.drill.exec.store.msgpack.valuewriter.BooleanValueWriter;
+import org.apache.drill.exec.store.msgpack.valuewriter.ExtensionValueWriter;
+import org.apache.drill.exec.store.msgpack.valuewriter.FloatValueWriter;
+import org.apache.drill.exec.store.msgpack.valuewriter.IntegerValueWriter;
+import org.apache.drill.exec.store.msgpack.valuewriter.MapValueWriter;
+import org.apache.drill.exec.store.msgpack.valuewriter.StringValueWriter;
+import org.apache.drill.exec.vector.complex.fn.FieldSelection;
+import org.apache.drill.exec.vector.complex.writer.BaseWriter;
+import org.apache.drill.exec.vector.complex.writer.BaseWriter.ComplexWriter;
+import org.apache.drill.exec.vector.complex.writer.BaseWriter.ListWriter;
+import org.apache.drill.shaded.guava.com.google.common.collect.Lists;
+import org.msgpack.core.MessageInsufficientBufferException;
+import org.msgpack.core.MessagePack;
+import org.msgpack.core.MessageUnpacker;
+import org.msgpack.value.MapValue;
+import org.msgpack.value.Value;
+import org.msgpack.value.ValueType;
+
+import io.netty.buffer.DrillBuf;
+
+public class MsgpackReader {
+
+  static final org.slf4j.Logger logger = org.slf4j.LoggerFactory.getLogger(MsgpackReader.class);
+  private final List<SchemaPath> columns;
+  private final FieldSelection rootSelection;
+  protected MessageUnpacker unpacker;
+  protected MsgpackReaderContext context;
+  private final boolean skipQuery;
+
+  /**
+   * Collection for tracking empty array writers during reading and storing them
+   * for initializing empty arrays
+   */
+  private final List<ListWriter> emptyArrayWriters = Lists.newArrayList();
+  private boolean allTextMode = false;
+  private MapValueWriter mapValueWriter;
+
+  public MsgpackReader(InputStream stream, MsgpackReaderContext context, DrillBuf managedBuf, List<SchemaPath> columns,
+      boolean skipQuery) {
+
+    this.context = context;
+    this.context.workBuf = managedBuf;
+    this.unpacker = MessagePack.newDefaultUnpacker(stream);
+    this.columns = columns;
+    this.skipQuery = skipQuery;
+    rootSelection = FieldSelection.getFieldSelection(columns);
+    EnumMap<ValueType, AbstractValueWriter> valueWriterMap = new EnumMap<>(ValueType.class);
+    valueWriterMap.put(ValueType.ARRAY, new ArrayValueWriter(valueWriterMap, emptyArrayWriters));
+    valueWriterMap.put(ValueType.FLOAT, new FloatValueWriter());
+    valueWriterMap.put(ValueType.INTEGER, new IntegerValueWriter());
+    valueWriterMap.put(ValueType.BOOLEAN, new BooleanValueWriter());
+    valueWriterMap.put(ValueType.STRING, new StringValueWriter());
+    valueWriterMap.put(ValueType.BINARY, new BinaryValueWriter());
+    valueWriterMap.put(ValueType.EXTENSION, new ExtensionValueWriter());
```

Comment:
Somewhat confused here; some comments may help.

Presumably, the file can contain any number of FLOAT fields, including 0. Each field needs its own state (reader, parser, whatever.) Each has its own value vector. How does it work to have one "ValueWriter" per type that takes no parameters to say which field or vector is being worked on?

Reply:
I'll put more comments in the code. Basically this is a switch implemented using an EnumMap. In the  ComplexValueWriter I use this switch to lookup what class will handle writing a value type. Here's the line of code from that writeElement method

      valueWriterMap.get(value.getValueType()).write(value, mapWriter, fieldName, listWriter, selection, schema);
So based on the value type I get the corresponding writer class to use.